### PR TITLE
feat(ekf_localizer): add diagnostics

### PR DESCRIPTION
**Pull request by YamatoAndo**
## Description

Add ekf_localizer's diagnostics.

ekf_diagnostics

Please refer to the readme for the conditions that trigger WARN/ERROR states.

## Related links

https://github.com/autowarefoundation/autoware_launch/pull/554

## Tests performed

Test with https://github.com/autowarefoundation/autoware_launch/pull/554

I have confirmed the following
- The logging simulator works with the sample rosbag
- The diagnostic state triggers WARN/ERROR under the conditions mentioned in the readme
- The unit test is passed

## Notes for reviewers

After merging this pr, I will modify [system_error_monitor.param.yaml](https://github.com/autowarefoundation/autoware.universe/blob/main/system/system_error_monitor/config/system_error_monitor.param.yaml) and [diagnostic_aggregator/localization.param.yaml](https://github.com/autowarefoundation/autoware.universe/blob/main/system/system_error_monitor/config/diagnostic_aggregator/localization.param.yaml) to notify ekf_localizer's state to the autoware system.

## Interface changes

-  add a new published topic 
   - `/diagnostics` 
- add parameters
  - `pose_no_update_count_threshold_warn`
  - `pose_no_update_count_threshold_error`
  - `twist_no_update_count_threshold_warn`
  - `twist_no_update_count_threshold_error`

## Effects on system behavior

 N/A

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Review by Motsu-san — CHANGES_REQUESTED** [review]
Big commit.
Could you separate the commit fc1dd2f97294fe0f137ad559396c3ddce22d3dc3 to feature changes, tests and docs(ex. README)?

**Comment by KYabuuchi** [comment]
Thank you for creating the pull request.  I have confirmed that the diagnostics works well in my environment too.  :+1:
Please wait a bit for the implementation review.

**Comment by YamatoAndo** [comment]
@SakodaShintaro 
This node is single-threaded, so parallel processing does not occur.
So I believe that checkProcessActivated will not be called in parallel while the destructor is running.

**Comment by YamatoAndo** [comment]
If my understanding is incorrect, please let me know.

**Comment by SakodaShintaro** [comment]
@YamatoAndo 
Oh, I see.  I understood that the diag is a one of single-threaded timer callback which runs on the same thread as ekf. Then it looks fine.

I would like to check the diagnostics working, but how did you get the image shown at the beginning of the Description?

**Comment by YamatoAndo** [comment]
@SakodaShintaro  I used following commnad to get that image.
`$ ros2 run rqt_runtime_monitor rqt_runtime_monitor`

**Comment by YamatoAndo** [comment]
Since there is a possibility of making significant changes, I will temporarily convert to draft.

**Comment by YamatoAndo** [comment]
@KYabuuchi @SakodaShintaro I have reopened the PR. Please review again.

**Review by SakodaShintaro — APPROVED** [review]
Looks Good To Me

File: localization/ekf_localizer/src/diagnostics.cpp [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  if (!is_passed_mahalanobis_gate) {
+    stat.level = diagnostic_msgs::msg::DiagnosticStatus::WARN;
+    stat.message = "[WARN]mahalanobis distance of " + measurement_type + " topic is large";
+  }
+
+  return stat;
+}
+
+// The highest level within the stat_array will be reflected in the merged_stat.
+// When all stat_array entries are 'OK,' the message of merged_stat will be empty
```
[the remark this answers is not in the document]
Comment (KYabuuchi) [reply localization/ekf_localizer/src/diagnostics.cpp:143]: When this comment is fixed I will approve the PR.

**Review by KYabuuchi — APPROVED** [review]
:+1:

**Comment by YamatoAndo** [comment]
@Motsu-san Could you please approve this pr?

**Review by kminoda — APPROVED** [review]
LGTM

**Review by Motsu-san — APPROVED** [review]
I approve this pull-request even without organizing the commits due to the circumstances of the squash merge rule. I hope that the pull-request will become smaller in size, which is a prerequisite for squash merge in the future.

**Comment by kminoda** [comment]
@Motsu-san Thank you!

**Pull request merged**